docs/optimization-tips.md

```diff
@@ -0,0 +1,30 @@
+Swift NIO Optimization Tips
+===========================
+
+The following document is a list of tips and advice for writing high performance
+code using Swift NIO.
+
+General Optimization Tips
+-------------------------
+
+Although not specific to NIO, the [Swift GitHub repository][swift-repo] has a
+list of tips and tricks for [writing high performance Swift code][swift-optimization-tips]
+and is a great place to start.
+
+Wrapping types in `NIOAny`
+--------------------------
+
+Anything passing through a NIO pipeline that is not special-cased by `NIOAny`
+(i.e. is not one of `ByteBuffer`, `FileRegion`, `IOData` or
+`AddressedEnvelope<ByteBuffer>`) needs to have careful attention paid to its
+size. If the type isn't one of the aforementioned special cases then it must be
+no greater than 24 bytes in size to avoid a heap-allocation.
```

Comment:
We should elaborate here.

"It must be no greater than 24 bytes in size to avoid a heap-allocation each time it is added to a `NIOAny`."

Reply:
You may also want to point to the use of `MemoryLayout.size` to work out the size for folks who are unsure.

Reply:
Good idea :)